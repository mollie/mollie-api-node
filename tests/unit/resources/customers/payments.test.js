import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import CustomersPayments from 'resources/customers/payments';
import Payment from 'models/payment';

import response from '../../__stubs__/customers_payments.json';

const mock = new MockAdapter(axios);

const props = {
  id: 'tr_McVBfE9Ceq',
  amount: {
    currency: 'EUR',
    value: '10.00',
  },
  description: 'Test customer payment of â‚¬10.00',
  customerId: 'cst_c24gk2t3G6',
  redirectUrl: 'https://www.example.org/',
};

describe('customers_payments', () => {
  let customersPayments;
  beforeEach(() => {
    customersPayments = new CustomersPayments(axios.create());
  });

  it('should have a resource name and model', () => {
    expect(CustomersPayments.resource).toBe('customers_payments');
    expect(CustomersPayments.model).toBe(Payment);
  });

  describe('.create()', () => {
    mock
      .onPost(`/customers/${props.customerId}/payments`)
      .reply(200, response._embedded.payments[0]);

    it('should return a payment instance', () =>
      customersPayments.create(props).then(result => {
        expect(result).toBeInstanceOf(Payment);
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      customersPayments.create(props, (err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Payment);
        expect(result).toMatchSnapshot();
        done();
      });
    });

    it('should work with withParent', done => {
      customersPayments
        .withParent({
          resource: 'customer',
          id: props.customerId,
        })
        .create(props, (err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Payment);
          expect(result).toMatchSnapshot();
          done();
        });
    });
  });

  describe('.get()', () => {
    const error = { error: { message: 'The customers_mandate id is invalid' } };

    mock
      .onGet(`/customers/${props.customerId}/payments/${props.id}`)
      .reply(200, response._embedded.payments[0]);
    mock.onGet(`/customers/${props.customerId}/payments/foo`).reply(500, error);

    it('should return a payment instance', () =>
      customersPayments.get(props.id, props).then(result => {
        expect(result).toBeInstanceOf(Payment);
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      customersPayments.get(props.id, props, (err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Payment);
        expect(result).toMatchSnapshot();
        done();
      });
    });

    it('should work with withParent', done => {
      customersPayments
        .withParent({
          resource: 'customer',
          id: props.customerId,
        })
        .get(props.id, (err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Payment);
          expect(result).toMatchSnapshot();
          done();
        });
    });
  });

  describe('.all()', () => {
    mock.onGet(`/customers/${props.customerId}/payments`).reply(200, response);

    it('should return a list of all customer payments', () =>
      customersPayments.all({ customerId: props.customerId }).then(result => {
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
      }));

    it('should throw an error if "customerId" is not set', () => {
      const getPayments = () => customersPayments.all();

      expect(getPayments).toThrowError(TypeError);
    });

    it('should work with a callback', done => {
      customersPayments.all(props, (err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
        done();
      });
    });

    it('should work with withParent', () =>
      customersPayments
        .withParent({
          resource: 'customer',
          id: props.customerId,
        })
        .all()
        .then(result => {
          expect(result).toBeInstanceOf(Array);
          expect(result).toHaveProperty('links');
          expect(result).toMatchSnapshot();
        }));
  });
});
