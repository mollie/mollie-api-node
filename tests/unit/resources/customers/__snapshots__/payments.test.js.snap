// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customers_payments .all() should return a list of all customer payments 1`] = `
Array [
  Payment {
    "_links": Object {
      "checkout": Object {
        "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "GBP",
      "value": "75.00",
    },
    "amountRefunded": null,
    "amountRemaining": null,
    "applicationFee": Object {
      "amount": Object {
        "currency": null,
        "value": null,
      },
      "description": null,
    },
    "canceledAt": null,
    "countryCode": null,
    "createdAt": "2018-02-12T11:58:35.0Z",
    "customerId": "cst_kEn1PlbGa",
    "description": "test",
    "details": null,
    "expiredAt": null,
    "expiresAt": "2018-02-12T12:13:35.0Z",
    "failedAt": null,
    "id": "tr_7UhSN1zuXS",
    "isCancelable": false,
    "locale": null,
    "mandateId": null,
    "metadata": null,
    "method": "ideal",
    "mode": "test",
    "paidAt": null,
    "profileId": "pfl_QkEhN94Ba",
    "redirectUrl": "https://webshop.example.org/order/12345/",
    "resource": "payment",
    "sequenceType": null,
    "settlementAmount": null,
    "settlementId": null,
    "status": "open",
    "subscriptionId": null,
    "webhookUrl": null,
  },
]
`;

exports[`customers_payments .all() should work with a callback 1`] = `
Array [
  Payment {
    "_links": Object {
      "checkout": Object {
        "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "GBP",
      "value": "75.00",
    },
    "amountRefunded": null,
    "amountRemaining": null,
    "applicationFee": Object {
      "amount": Object {
        "currency": null,
        "value": null,
      },
      "description": null,
    },
    "canceledAt": null,
    "countryCode": null,
    "createdAt": "2018-02-12T11:58:35.0Z",
    "customerId": "cst_kEn1PlbGa",
    "description": "test",
    "details": null,
    "expiredAt": null,
    "expiresAt": "2018-02-12T12:13:35.0Z",
    "failedAt": null,
    "id": "tr_7UhSN1zuXS",
    "isCancelable": false,
    "locale": null,
    "mandateId": null,
    "metadata": null,
    "method": "ideal",
    "mode": "test",
    "paidAt": null,
    "profileId": "pfl_QkEhN94Ba",
    "redirectUrl": "https://webshop.example.org/order/12345/",
    "resource": "payment",
    "sequenceType": null,
    "settlementAmount": null,
    "settlementId": null,
    "status": "open",
    "subscriptionId": null,
    "webhookUrl": null,
  },
]
`;

exports[`customers_payments .all() should work with withParent 1`] = `
Array [
  Payment {
    "_links": Object {
      "checkout": Object {
        "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "GBP",
      "value": "75.00",
    },
    "amountRefunded": null,
    "amountRemaining": null,
    "applicationFee": Object {
      "amount": Object {
        "currency": null,
        "value": null,
      },
      "description": null,
    },
    "canceledAt": null,
    "countryCode": null,
    "createdAt": "2018-02-12T11:58:35.0Z",
    "customerId": "cst_kEn1PlbGa",
    "description": "test",
    "details": null,
    "expiredAt": null,
    "expiresAt": "2018-02-12T12:13:35.0Z",
    "failedAt": null,
    "id": "tr_7UhSN1zuXS",
    "isCancelable": false,
    "locale": null,
    "mandateId": null,
    "metadata": null,
    "method": "ideal",
    "mode": "test",
    "paidAt": null,
    "profileId": "pfl_QkEhN94Ba",
    "redirectUrl": "https://webshop.example.org/order/12345/",
    "resource": "payment",
    "sequenceType": null,
    "settlementAmount": null,
    "settlementId": null,
    "status": "open",
    "subscriptionId": null,
    "webhookUrl": null,
  },
]
`;

exports[`customers_payments .create() should return a payment instance 1`] = `
Payment {
  "_links": Object {
    "checkout": Object {
      "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
      "type": "text/html",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "GBP",
    "value": "75.00",
  },
  "amountRefunded": null,
  "amountRemaining": null,
  "applicationFee": Object {
    "amount": Object {
      "currency": null,
      "value": null,
    },
    "description": null,
  },
  "canceledAt": null,
  "countryCode": null,
  "createdAt": "2018-02-12T11:58:35.0Z",
  "customerId": "cst_kEn1PlbGa",
  "description": "test",
  "details": null,
  "expiredAt": null,
  "expiresAt": "2018-02-12T12:13:35.0Z",
  "failedAt": null,
  "id": "tr_7UhSN1zuXS",
  "isCancelable": false,
  "locale": null,
  "mandateId": null,
  "metadata": null,
  "method": "ideal",
  "mode": "test",
  "paidAt": null,
  "profileId": "pfl_QkEhN94Ba",
  "redirectUrl": "https://webshop.example.org/order/12345/",
  "resource": "payment",
  "sequenceType": null,
  "settlementAmount": null,
  "settlementId": null,
  "status": "open",
  "subscriptionId": null,
  "webhookUrl": null,
}
`;

exports[`customers_payments .create() should work with a callback 1`] = `
Payment {
  "_links": Object {
    "checkout": Object {
      "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
      "type": "text/html",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "GBP",
    "value": "75.00",
  },
  "amountRefunded": null,
  "amountRemaining": null,
  "applicationFee": Object {
    "amount": Object {
      "currency": null,
      "value": null,
    },
    "description": null,
  },
  "canceledAt": null,
  "countryCode": null,
  "createdAt": "2018-02-12T11:58:35.0Z",
  "customerId": "cst_kEn1PlbGa",
  "description": "test",
  "details": null,
  "expiredAt": null,
  "expiresAt": "2018-02-12T12:13:35.0Z",
  "failedAt": null,
  "id": "tr_7UhSN1zuXS",
  "isCancelable": false,
  "locale": null,
  "mandateId": null,
  "metadata": null,
  "method": "ideal",
  "mode": "test",
  "paidAt": null,
  "profileId": "pfl_QkEhN94Ba",
  "redirectUrl": "https://webshop.example.org/order/12345/",
  "resource": "payment",
  "sequenceType": null,
  "settlementAmount": null,
  "settlementId": null,
  "status": "open",
  "subscriptionId": null,
  "webhookUrl": null,
}
`;

exports[`customers_payments .create() should work with withParent 1`] = `
Payment {
  "_links": Object {
    "checkout": Object {
      "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
      "type": "text/html",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "GBP",
    "value": "75.00",
  },
  "amountRefunded": null,
  "amountRemaining": null,
  "applicationFee": Object {
    "amount": Object {
      "currency": null,
      "value": null,
    },
    "description": null,
  },
  "canceledAt": null,
  "countryCode": null,
  "createdAt": "2018-02-12T11:58:35.0Z",
  "customerId": "cst_kEn1PlbGa",
  "description": "test",
  "details": null,
  "expiredAt": null,
  "expiresAt": "2018-02-12T12:13:35.0Z",
  "failedAt": null,
  "id": "tr_7UhSN1zuXS",
  "isCancelable": false,
  "locale": null,
  "mandateId": null,
  "metadata": null,
  "method": "ideal",
  "mode": "test",
  "paidAt": null,
  "profileId": "pfl_QkEhN94Ba",
  "redirectUrl": "https://webshop.example.org/order/12345/",
  "resource": "payment",
  "sequenceType": null,
  "settlementAmount": null,
  "settlementId": null,
  "status": "open",
  "subscriptionId": null,
  "webhookUrl": null,
}
`;

exports[`customers_payments .get() should return a payment instance 1`] = `
Payment {
  "_links": Object {
    "checkout": Object {
      "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
      "type": "text/html",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "GBP",
    "value": "75.00",
  },
  "amountRefunded": null,
  "amountRemaining": null,
  "applicationFee": Object {
    "amount": Object {
      "currency": null,
      "value": null,
    },
    "description": null,
  },
  "canceledAt": null,
  "countryCode": null,
  "createdAt": "2018-02-12T11:58:35.0Z",
  "customerId": "cst_kEn1PlbGa",
  "description": "test",
  "details": null,
  "expiredAt": null,
  "expiresAt": "2018-02-12T12:13:35.0Z",
  "failedAt": null,
  "id": "tr_7UhSN1zuXS",
  "isCancelable": false,
  "locale": null,
  "mandateId": null,
  "metadata": null,
  "method": "ideal",
  "mode": "test",
  "paidAt": null,
  "profileId": "pfl_QkEhN94Ba",
  "redirectUrl": "https://webshop.example.org/order/12345/",
  "resource": "payment",
  "sequenceType": null,
  "settlementAmount": null,
  "settlementId": null,
  "status": "open",
  "subscriptionId": null,
  "webhookUrl": null,
}
`;

exports[`customers_payments .get() should work with a callback 1`] = `
Payment {
  "_links": Object {
    "checkout": Object {
      "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
      "type": "text/html",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "GBP",
    "value": "75.00",
  },
  "amountRefunded": null,
  "amountRemaining": null,
  "applicationFee": Object {
    "amount": Object {
      "currency": null,
      "value": null,
    },
    "description": null,
  },
  "canceledAt": null,
  "countryCode": null,
  "createdAt": "2018-02-12T11:58:35.0Z",
  "customerId": "cst_kEn1PlbGa",
  "description": "test",
  "details": null,
  "expiredAt": null,
  "expiresAt": "2018-02-12T12:13:35.0Z",
  "failedAt": null,
  "id": "tr_7UhSN1zuXS",
  "isCancelable": false,
  "locale": null,
  "mandateId": null,
  "metadata": null,
  "method": "ideal",
  "mode": "test",
  "paidAt": null,
  "profileId": "pfl_QkEhN94Ba",
  "redirectUrl": "https://webshop.example.org/order/12345/",
  "resource": "payment",
  "sequenceType": null,
  "settlementAmount": null,
  "settlementId": null,
  "status": "open",
  "subscriptionId": null,
  "webhookUrl": null,
}
`;

exports[`customers_payments .get() should work with withParent 1`] = `
Payment {
  "_links": Object {
    "checkout": Object {
      "href": "https://www.mollie.com/paymentscreen/issuer/select/ideal/7UhSN1zuXS",
      "type": "text/html",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/payments/tr_7UhSN1zuXS",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "GBP",
    "value": "75.00",
  },
  "amountRefunded": null,
  "amountRemaining": null,
  "applicationFee": Object {
    "amount": Object {
      "currency": null,
      "value": null,
    },
    "description": null,
  },
  "canceledAt": null,
  "countryCode": null,
  "createdAt": "2018-02-12T11:58:35.0Z",
  "customerId": "cst_kEn1PlbGa",
  "description": "test",
  "details": null,
  "expiredAt": null,
  "expiresAt": "2018-02-12T12:13:35.0Z",
  "failedAt": null,
  "id": "tr_7UhSN1zuXS",
  "isCancelable": false,
  "locale": null,
  "mandateId": null,
  "metadata": null,
  "method": "ideal",
  "mode": "test",
  "paidAt": null,
  "profileId": "pfl_QkEhN94Ba",
  "redirectUrl": "https://webshop.example.org/order/12345/",
  "resource": "payment",
  "sequenceType": null,
  "settlementAmount": null,
  "settlementId": null,
  "status": "open",
  "subscriptionId": null,
  "webhookUrl": null,
}
`;
